### DISCUSSION TITLE
Get Handle to existing Asset (from the AssetId)

### DISCUSSION THREAD

**Discussion by thmxv** [bevyengine/bevy#11743] [Q&A] [answered]
Hi,

I build Asset from code and add them using Assets<A>.add(). However, I would like to search for possible duplicate and get a handle to the duplicate, if it exists, before doing the add.

```rust
let new_asset = MyAssetType {
    data: 1,
};
// TODO: Do not call following code if asset is a duplicate of an existing one
// set the handle to the pre-existing asset
let asset_handle = my_assets.add(new_asset);
```

I spent a lot of time reading the doc and trying some code. I can find duplicates with:

```rust
my_asserts.iter().find(|(id, a)| **a == new_asset)
```
But I do not manage to get a strong handle to the existing asset.

Am I doing something wrong? Is there way to do what I want?

**Comment by thmxv** [comment]
After experimenting and looking more in the doc, I found something that works. After searching the assets to find an equality and get the ID of the matching asset, I can use [AssetServer::get_id_handle](https://docs.rs/bevy/latest/bevy/asset/struct.AssetServer.html#method.get_id_handle) to get a strong handle to the corresponding asset.

**Comment by thmxv** [comment]
No luck. I tried to use what I found but it crashes when unwrapping the Option returned by AssetServer::get_id_handle(). I can avoid the crash by adding the asset to the AssetServer instead of to the Assets but, in this case, the lookup for already existing assets does not work until the next frame.

```rust
pub fn get_existing_or_add_asset<A, P>(
    asset: A,
    assets: &mut Assets<A>,
    asset_server: &AssetServer,
    predicate: P,
) -> Handle<A>
where
    A: Asset,
    P: Fn(&A, &A) -> bool,
{
    let existing_asset = assets.iter().find(|(_id, a)| predicate(*a, &asset));
    let handle = match existing_asset {
        Some((id, _a)) => {
            println!("Asset found");
            // FIXME: Crash into unwrap
            asset_server.get_id_handle(id).unwrap()
        }
        None => {
            assets.add(asset)
            // asset_server.add(asset)
        }
    };
    assert!(handle.is_strong());
    handle
}
```
I suppose the only solution (I think of) is to use another Resource (of type HashMap::<AssetID, Handle>) that maps the AssetId to a weak Handle. And if the weak Handle points to a Asset that is still alive return a Strong clone of this handle, if the Asset is not still alive than adding it to the Assets and returning the resulting Handle (and storing a weak clone into the map).

Any help is welcome. I do not know if the solution I think about would work and I think there must be a way (unknown to me) to get an Handle from an AssetId that do not involve the AssetServer (just the Assets) but I cannot find any (after spending days looking in the doc and the code). 

Thanks for the engine anyways, still a great library even if some basic things like this makes it a bit hard for a newcomer to rust and this engine.

**Reply by nicopap** [reply, marked as answer]
Just a tip, in the rust doc, you can search for "function that accept type `X` and return type `Y`" with the `->`. As follow: <https://docs.rs/bevy/latest/bevy/asset/enum.AssetId.html?search=AssetId%20-%3E%20Handle%3CA%3E>.

I did investigate the `get_id_handle` implementation and it looks like a bug to me. I'll open an issue.

**Reply by nicopap** [reply]
Yeah, I think that storing the strong handle somewhere and retrieving it is currently the best approach.

Personally I would have gone with a `HashMap<A, Handle<A>>`. If `A` has a lot of data, comparing each individual `A` in the `Assets<A>` is going to run a lot of CPU cycles. While with a `HashMap`, you basically compare against a single `A`, as it compares the hash before (which is very small), then compare the value.

**Reply by thmxv** [reply]
Wow. Thank you so much, this is much more than I was expecting. The answer, the issue report and the related PR is a lot of work useful to me.

Regarding the map to use as a workaround and to better performances: Using an `Asset` as the key would make the `HashMap` own a copy of the `Asset` and introduce a duplicate which is what I am trying to avoid. Because I am not working on a game but on a DB application (which needs to be quick and display the results of DB queries in 3D) and the concerned `Asset`s are corresponding to a DB record (a row of a table in the DB), I can use a `HashMap<u32, Handle<A>>` to map the DB ID to the `Handle`. I can even store the ID in the `Component`s (instead of a `Handle<A>`), map the ID to the DB model struct (using a `HashMap<u32, RowDbModelStruct>`,  and only pay the price of the growing memory consumption due to not unloading the assets when there are not used by any `Component` anymore.
